cmd/internal/kube/daemonset/daemonset.go

```diff
@@ -135,48 +227,143 @@ func (d *Daemonset) KubernetesEventsWorker() {
 	// Not implemented
 }
 
-func (d *Daemonset) PodsHealthWorker() {
-	// Not implemented
+func (d *Daemonset) checkPodsHealth() {
+	var podNames []string
+	d.PodArgsByNameMap.Range(func(k, v interface{}) bool {
+		e := v.(*PodArgs)
+		podNames = append(podNames, e.PodName)
+		return true
+	})
+
+	podStatuses, err := d.KubeClient.GetPodsStatus(podNames)
+	if err != nil {
+		printer.Errorf("failed to get pods status: %v", err)
+		return
+	}
+
+	for podName, podStatus := range podStatuses {
+		if podStatus == string(coreV1.PodSucceeded) || podStatus == string(coreV1.PodFailed) {
+			printer.Infof("pod %s has stopped running", podStatus)
+
+			podArgs, err := d.getPodArgsFromMap(podName)
+			if err != nil {
+				printer.Errorf("failed to get podArgs for pod %s: %v", podName, err)
+				continue
+			}
+
+			isSameStage, err := podArgs.validatePodTrafficMonitorStage(PodTerminated, TrafficMonitoringStarted)
```

Comment:
Yeah, mutex is not held here -- should we grab it around all the interaction with the podArgs object?  I would expect it's easier to reason about

get mutex
validate new state
set state
release mutex

Then about acquiring and releasing the mutex twice, which means an extra transition could sneak in.